source/fe/mapping_q1.cc

```diff
@@ -92,6 +93,246 @@ namespace internal
+          const double x20 = -x18 + x19*y0 - x19*y2 + xi*y3;
+          const double x21 = xi*y5;
+          const double x22 = eta*xi;
+          const double x23 = -eta;
+          const double x24 = x22 + x23;
+          const double x25 = x10*xi + x23 + 1;
+          const double x26 = -eta*z2 + eta*z3 + x10*z0 - x10*z1;
+          const double x27 = xi*z1;
+          const double x28 = x19*z0 - x19*z2 - x27 + xi*z3;
+          const double x29 = xi*z5;
+          const double x30 = zeta - 1;
+
+          const double j00 = x11 + x12 - x14*zeta + x8 + x9;
+          const double j01 = x15 + x16*xi
+                             - zeta*(x15 + x4 - x6 + xi*(x16 - x4 + x5 + x6 - x7));
+          const double j02 = x11 + x13 - x14*xi + x9;
+          const double j10 = x17 - zeta*(eta*y6 - eta*y7 - x10*y4 + x10*y5 + x17);
+          const double j11 = x20 - zeta*(-x19*y4 + x19*y6 + x20 + x21 - xi*y7);
+          const double j12 = x10*x18 - x10*x21 - x22*y3 + x22*y7 + x24*y2
+                             - x24*y6 - x25*y0 + x25*y4;
+          const double j20 = x26 - zeta*(eta*z6 - eta*z7 - x10*z4 + x10*z5 + x26);
+          const double j21 = x28 - zeta*(-x19*z4 + x19*z6 + x28 + x29 - xi*z7);
+          const double j22 = x10*x27 - x10*x29 - x22*z3 + x22*z7 + x24*z2
+                             - x24*z6 - x25*z0 + x25*z4;
+
+          const double det = -(j12*j21 - j11*j22)*j00 + (j02*j21 - j01*j22)*j10
+                             - (j02*j11 - j01*j12)*j20;
+          const double jinv00 = (-j12*j21 + j11*j22)/det;
+          const double jinv01 = (j02*j21 - j01*j22)/det;
+          const double jinv02 = (-j02*j11 + j01*j12)/det;
+          const double jinv10 = (j12*j20 - j10*j22)/det;
+          const double jinv11 = (-j02*j20 + j00*j22)/det;
+          const double jinv12 = (j02*j10 - j00*j12)/det;
+          const double jinv20 = (-j11*j20 + j10*j21)/det;
+          const double jinv21 = (j01*j20 - j00*j21)/det;
+          const double jinv22 = (-j01*j10 + j00*j11)/det;
+
+          const double f0 = -((x0*x19 - x1*xi)*(x10) - (x2*(x19) - x3*xi)*eta)*x30
+                            + ((x4*x19 - x5*xi)*x10 - (x6*x19 - x7*xi)*eta)*zeta - x;
+          const double f1 = -((x19*y0 - xi*y1)*x10 - (x19*y2 - xi*y3)*eta)*x30
+                            + ((x19*y4 - xi*y5)*x10 - (x19*y6 - xi*y7)*eta)*zeta - y;
+          const double f2 = -((x19*z0 - xi*z1)*x10 - (x19*z2 - xi*z3)*eta)*x30
+                            + ((x19*z4 - xi*z5)*x10 - (x19*z6 - xi*z7)*eta)*zeta - z;
+
+          const double update0 = jinv00*f0 + jinv01*f1 + jinv02*f2;
+          const double update1 = jinv10*f0 + jinv11*f1 + jinv12*f2;
+          const double update2 = jinv20*f0 + jinv21*f1 + jinv22*f2;
+          xi -= update0;
+          eta -= update1;
+          zeta -= update2;
+          if ((std::abs(update0) < 1e-12) && (std::abs(update1) < 1e-12)
+              && (std::abs(update2) < 1e-12))
+            {
+              break;
+            }
+
```

Comment:
This is a classical Newton algorithm for finding the point location. Please add some comment at the top of this loop.